fuel-tx/src/tests/da_compression.rs

```diff
@@ -0,0 +1,242 @@
+use fuel_compression::{
+    CompressibleBy,
+    DecompressibleBy,
+    RegistryKey,
+};
+use fuel_crypto::SecretKey;
+use fuel_types::{
+    Address,
+    AssetId,
+    ContractId,
+};
+use rand::{
+    rngs::StdRng,
+    Rng,
+    SeedableRng,
+};
+use std::{
+    collections::HashMap,
+    convert::Infallible,
+};
+
+type Keyspace = &'static str;
+
+/// A simple and inefficient registry for testing purposes
+#[derive(Default)]
+struct TestCompressionCtx {
+    registry: HashMap<Keyspace, BiMap<RegistryKey, Vec<u8>>>,
+    tx_blocks: BiMap<TxPointer, CompressibleTxId>,
+}
+
+macro_rules! impl_substitutable_key {
+    ($t:ty) => {
+        impl CompressibleBy<TestCompressionCtx, Infallible> for $t {
+            async fn compress(
+                &self,
+                ctx: &mut TestCompressionCtx,
+            ) -> Result<RegistryKey, Infallible> {
+                let keyspace = stringify!($t);
+                let value = postcard::to_stdvec(self).expect("failed to serialize");
+                let key_seed = ctx.registry.len(); // Just get an unique integer key
+
+                let entry = ctx.registry.entry(keyspace).or_default();
+                if let Some(key) = entry.get_by_right(&value) {
+                    return Ok(*key);
+                }
+
+                let key = RegistryKey::try_from(key_seed as u32).expect("key too large");
+                entry.insert(key, value);
+                Ok(key)
+            }
+        }
+
+        impl DecompressibleBy<TestCompressionCtx, Infallible> for $t {
+            async fn decompress(
+                key: &RegistryKey,
+                ctx: &TestCompressionCtx,
+            ) -> Result<$t, Infallible> {
+                let keyspace = stringify!($t);
+                let values = ctx.registry.get(&keyspace).expect("key not found");
+                let value = values.get_by_left(key).expect("key not found");
+                Ok(postcard::from_bytes(value).expect("failed to deserialize"))
+            }
+        }
+    };
+}
+
+impl_substitutable_key!(Address);
+impl_substitutable_key!(AssetId);
+impl_substitutable_key!(ContractId);
+impl_substitutable_key!(ScriptCode);
+
+impl CompressibleBy<TestCompressionCtx, Infallible> for CompressibleTxId {
+    async fn compress(
+        &self,
+        ctx: &mut TestCompressionCtx,
+    ) -> Result<TxPointer, Infallible> {
+        if let Some(key) = ctx.tx_blocks.get_by_right(self) {
+            return Ok(*key);
+        }
+
+        let key_seed = ctx.tx_blocks.len(); // Just get an unique integer key
+        let key = TxPointer::new((key_seed as u32).into(), 0);
+        ctx.tx_blocks.insert(key, *self);
+        Ok(key)
+    }
+}
+
+impl DecompressibleBy<TestCompressionCtx, Infallible> for CompressibleTxId {
+    async fn decompress(
+        key: &TxPointer,
+        ctx: &TestCompressionCtx,
+    ) -> Result<CompressibleTxId, Infallible> {
+        Ok(*ctx.tx_blocks.get_by_left(key).expect("key not found"))
+    }
+}
+
```

Comment:
It would be nice to test full compression and decompression with non zero values of the transaction with different inputs and outputs and check that results equal. We could reuse `TransactionFactory` to generate random transactions=)

Reply:
I'm a bit worried about the `TransactionFactory` not producing all kinds of transactions. This code fails for some case, but when I use `TransactionFactory` all txs pass.

Reply:
Added the full test in 89654f1f